sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/AddFields.java

```diff
@@ -0,0 +1,432 @@
+      this.newFields = newFields;
+    }
+
+    /**
+     * Add a new field of the specified type. The new field will be nullable and will be filled in
+     * with null values.
+     */
+    public Inner<T> field(String fieldName, Schema.FieldType fieldType) {
+      return field(fieldName, fieldType.withNullable(true), null);
+    }
+
+    /**
+     * Add a new field of the specified type. The new field will be filled in with the specified
+     * value.
+     */
+    public Inner<T> field(String fieldName, Schema.FieldType fieldType, Object defaultValue) {
+      if (defaultValue == null) {
+        checkArgument(fieldType.getNullable());
+      }
+
+      FieldAccessDescriptor descriptor = FieldAccessDescriptor.withFieldNames(fieldName);
+      checkArgument(descriptor.referencesSingleField());
+      List<NewField> fields =
+          ImmutableList.<NewField>builder()
+              .addAll(newFields)
+              .add(NewField.of(descriptor, fieldType, defaultValue))
+              .build();
+      return new Inner<>(fields);
+    }
+
+    private AddFieldsInformation getAddFieldsInformation(
+        Schema inputSchema, Collection<NewField> fieldsToAdd) {
+      List<NewField> newTopLevelFields =
+          fieldsToAdd.stream()
+              .filter(n -> !n.getDescriptor().getFieldsAccessed().isEmpty())
+              .collect(Collectors.toList());
+      List<NewField> newNestedFields =
+          fieldsToAdd.stream()
+              .filter(n -> !n.getDescriptor().getNestedFieldsAccessed().isEmpty())
+              .collect(Collectors.toList());
+      // Group all nested fields together by the field at the current level. For example, if adding
+      // a.b, a.c, a.d
+      // this map will contain a -> {a.b, a.c, a.d}/
+      Multimap<String, NewField> newNestedFieldsMap =
+          Multimaps.index(newNestedFields, NewField::getName);
+
+      Map<Integer, AddFieldsInformation> resolvedNestedNewValues = Maps.newHashMap();
+      Schema.Builder builder = Schema.builder();
+      for (int i = 0; i < inputSchema.getFieldCount(); ++i) {
+        Schema.Field field = inputSchema.getField(i);
+        Collection<NewField> nestedFields = newNestedFieldsMap.get(field.getName());
+
+        // If this field is a nested field and new subfields are added further down the tree, add
+        // those subfields before
+        // adding to the current schema. Otherwise we just add this field as is to the new schema.
+        if (!nestedFields.isEmpty()) {
+          nestedFields = nestedFields.stream().map(NewField::descend).collect(Collectors.toList());
+
+          AddFieldsInformation nestedInformation =
+              getAddFieldsInformation(field.getType(), nestedFields);
+          field = field.withType(nestedInformation.getOutputFieldType());
+          resolvedNestedNewValues.put(i, nestedInformation);
+        }
+        builder.addField(field);
+      }
+
+      // Add any new fields at this level.
```

Comment:
By reading the implementation code here, I see that top-level fields are added before nested fields. Can we document this in the class level javadoc? Otherwise this may be confusing to users, e.g. after calling
`pc.apply(AddFields.create().field("new1.nested", FieldType.STRING).field("new2", FieldType.INT32)`
users may expect a new schema in which `new1` appears before `new2`, which is not the case.

Reply:
Will talk about this more generally below.